### PR TITLE
[Issue 188] Architecture diagram

### LINKED ISSUE #188: [Story]: Architecture Diagram
### Summary

Create an initial architecture diagram for the backend and frontend that will serve as a living document in the main repository. 

This architecture diagram will allow: 
- stakeholders to understand the overall system design at a glance
- the team to make complex concepts and relationships easier to understand
- the team to show the system dependencies, interfaces, and interactions which will allow stakeholders to evaluate impact of design choices, identify risks, make informed decisions regarding system modifications, scalability, and future enhancements. 
- Use a tool that is version controlled, like mermaid.js
- We would need host diagram or application diagram

Examples of a good place to start for level of depth: 
- https://github.com/usds/justice40-tool/blob/main/docs/architecture/architecture.mmd
- https://github.com/usds/justice40-tool/tree/main/docs/architecture

### User story

As a software architect, I want to create an initial architecture diagram for the backend so that I can have a visual representation of the system's structure, components, and their relationships. 

### Acceptance criteria

- [ ] Create a living document to visually represent the architecture
- [ ] Following individuals have approved: 
  - @acouch
  - engineer not working the ticket @daphnegold @SammySteiner
  - @lucasmbrown-usds 
  - @widal001 
- [ ] Following individuals have been notified for option review and feedback
  - @gcarson-bit
  - @sumiat
- [ ] Diagram is created and stored in //TBD//
- [ ] Secrets manager is defined in the architecture diagram

### Describe the proposed testing strategy

_No response_

### Screenshots and additional context

_No response_

Blocked by #222

### PR DESCRIPTION
**Pull request by SammySteiner**
## Summary
Fixes #188 

### Time to review: 30

> It would be really helpful to get feedback on both the diagrams themselves and what they are missing, as well as my implementation of mermaid because I'm sure I have a lot to learn there.

### CODE REVIEW

**Comment by SammySteiner** [comment]
Apparently the font-awesome icons don't show up in github 🤦 
https://mermaid.js.org/syntax/flowchart.html#basic-support-for-fontawesome

**Comment by SammySteiner** [comment]
> 

Good question, I wasn't sure what to do with that. I imagine while we're using the replica database living in the microhealth tenant, those are definitely separate. But once we set up replication, I wasn't sure if we would keep the logging and user access tables for our api in a separate db or just in separate tables in the replicated database.

File: documentation/architecture/README.md [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+            Elastic Container Service"]:::ecs
+            ECR --> ECSS
+        end
+        GH --Build and deploys image--> ECR
+        GH --Restarts task with new image--> ECSS
+
+        classDef ecs fill:#FF9900,color:black
+```
+
+## Relevant ADRs
```
Comment (widal001) [inline documentation/architecture/README.md:257]: Love the links to the relevant ADRs!!! Great addition.

File: documentation/architecture/README.md [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+        Container n"]
+    end
+
+    subgraph RDS
+        subgraph private-subnet3 [Private Subnet]
+            DB[("Multi-AZ Grants.gov
+            Replica Database
+            tstgrntsee1.cshqswcusbcd.
+            us-east-1.rds.amazonaws.com
+            :1521")]:::db
```
Comment (widal001) [inline documentation/architecture/README.md:98]: I know this is in a private subnet so there shouldn't necessarily be a security concern, but my instinct would be avoid hard coding IP addresses and ports into the diagram -- unless you feel like including those details here provides important system-design-level information.

File: documentation/architecture/README.md [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+            PDB[("Multi-AZ Postgres DB
+            for Back End
+            IP.TBD:5432")]:::db
```
Comment (widal001) [inline documentation/architecture/README.md:101]: Same comment as above, I'd exclude this from the final diagram.

File: documentation/architecture/README.md [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        Container 1"]
+        c8[" Docker
+        Container n"]
+    end
+
+    subgraph RDS
+        subgraph private-subnet3 [Private Subnet]
+            DB[("Multi-AZ Grants.gov
+            Replica Database
+            tstgrntsee1.cshqswcusbcd.
```
Comment (widal001) [inline documentation/architecture/README.md:203]: Same comment above re: IP address and Port details

**Review by widal001 — COMMENTED** [review]
This looks awesome!! I left a few comments about some of the details hard coded in the diagram that we can discuss further in slack. 

Once we reach alignment on that question, I'll approve!

File: documentation/architecture/README.md (hunk 2 above)
Comment (acouch) [reply documentation/architecture/README.md:98]: I would agree w/ removing the actual address.

**Comment by acouch** [comment]
Looks great! Along with the comment above, I would also either remove the "IAM," "SSM," and "Cloudwatch" or integrate them into the flow. Does cloudwatch trigger the container update, or is it just for logging? If it is the latter, pulling it out into its own graph could help indicate that 
image

**Review by acouch — CHANGES_REQUESTED** [review]
Looks great! Along with the comment above, I would also either remove the "IAM," "SSM," and "Cloudwatch" or integrate them into the flow. Does cloudwatch trigger the container update, or is it just for logging? If it is the latter, pulling it out into its own graph could help indicate that 
image

**Comment by lucasmbrown-usds** [comment]
A couple stray thoughts: 

1. Will the front-end point at the API endpoints of the backend through the public internet (i.e., there's no privileged access of the front-end)? 

2. To convert data from the existing data format (in the replica DB) into a streamlined, new data format (in the Postgres DB), there's probably some type of conversion task (maybe it's stream processing where changes in the replica trigger a task to update data in the new Postgres DB? or maybe batch processing). Is that represented here as part of the ECS Back-End Tasks, or do we want to call it out separately? 

Otherwise looks great!

**Comment by acouch** [comment]
> Will the front-end point at the API endpoints of the backend through the public internet (i.e., there's no privileged access of the front-end)?

At some point. I would imagine that the need for HHS internal users to edit the text would be the driver there. We can update the diagram once we make that change.

 > To convert data from the existing data format (in the replica DB) into a streamlined, new data format (in the Postgres DB), there's probably some type of conversion task (maybe it's stream processing where changes in the replica trigger a task to update data in the new Postgres DB? or maybe batch processing). Is that represented here as part of the ECS Back-End Tasks, or do we want to call it out separately?
 
We are still doing discovery (#369 ) on what service or set of tools we will do that with. I think the replication could be a separate diagram as you've noted, once we have a strategy in place. I do think we could be more explicit about that relationship in this diagram.

**Comment by SammySteiner** [comment]
> 

I'd like to keep those in the general architecture as one of the goals is to document all the AWS services we're using, but I like the idea of subgraphing them to identify how they are being used.

**Review by acouch — APPROVED** [review]
Looks fantastic!

**Pull request merged**